### PR TITLE
[TKW] Make shared mem read/writes using non masked ops

### PR DESCRIPTION
**Pull request by Hardcode84**
Add a pass which adjusts shared mem reads/writes and MMA indices to never had to partial accesses. This is need to remove masking ops on shared mem.

Also, update `minimize_global_loads.py` to use ceildiv instead if floor div in case of unaligned sizes.

### CODE REVIEW

File: iree/turbine/kernel/wave/utils.py [hunk 1]
```diff
@@ -262,21 +262,44 @@ def get_hardware_vector_map(constraints: list[Constraint]) -> dict[IndexSymbol,
...
+    This function aligns index vars with Workgroup/Tiling constraints so it never
+    need partial reads/writes.
+    """
+    key_subs = {
+        c.dim: (c.count * c.tile_size)
+        for c in constraints
+        if isinstance(c, (TilingConstraint, WorkgroupConstraint))
+        and subs_idxc(c.dim) != subs_idxc(c.count * c.tile_size)
+    }
+    return {safe_subs(key, key_subs): index[key] for key in index}
```
Comment (harsh-nod) [inline iree/turbine/kernel/wave/utils.py:300]: So this is interesting. Stepping through the code, it seems like the original index 
```
{M: $T0 + BLOCK_M*floor($T0/64), N: $T1*BLOCK_N + 4*$T1 : 4 : 1}
```
would get transformed to 
```
{BLOCK_M*ceiling(M/BLOCK_M): $T0 + BLOCK_M*floor($T0/64), BLOCK_N*ceiling(N/BLOCK_N): $T1*BLOCK_N + 4*$T1 : 4 : 1}
```
But the problem with this is that, downstream passes and transformations can no longer query the original dimensions "M" and "N". Is there a way we can avoid modifying the dimension names?

Comment (Hardcode84) [reply iree/turbine/kernel/wave/utils.py:300]: Yes, this is true, and it was causing problems in thread shape analysis. And that's why I'm doing it almost at the very end of the pipeline.

Comment (harsh-nod) [reply iree/turbine/kernel/wave/utils.py:300]: I think that makes sense, but I am wondering whether it is necessary. So why do you need to rewrite M -> BLOCK_M * ceiling(M / BLOCK_M) ? How is that used in the codegen?

Comment (harsh-nod) [reply iree/turbine/kernel/wave/utils.py:300]: Couldn't you modify the value of M in the IndexingContext ?

Comment (Hardcode84) [reply iree/turbine/kernel/wave/utils.py:300]: `find_index_bounds` compares dim with aligned WG/Tilied size to determine if op need masking. We need original dim value to make making on global mem.

Comment (Hardcode84) [reply iree/turbine/kernel/wave/utils.py:300]: added a comment

**Review by harsh-nod — APPROVED** [review]
I would like to explore other ways to surface this information to earlier transformations, but this is good for now. thanks!

**Pull request merged**